src/Microsoft.AspNet.Server.Kestrel/Http/Frame.cs

```diff
@@ -317,6 +317,31 @@ public void ProduceStart()
 
         public void ProduceEnd(Exception ex)
         {
+            if (ex != null)
+            {
+                if (_resultStarted)
+                {
+                    // We can no longer respond with a 500, so we simply close the connection.
+                    ConnectionControl.End(ProduceEndType.SocketDisconnect);
+                    return;
+                }
+                else
+                {
+                    StatusCode = 500;
+                    ReasonPhrase = null;
+
+                    // If OnStarting hasn't been triggered yet, we don't want to trigger it now that
+                    // the app func has failed. https://github.com/aspnet/KestrelHttpServer/issues/43
+                    _onStarting = null;
+
+                    ResponseHeaders.Clear();
+                    if (_keepAlive)
+                    {
+                        ResponseHeaders["Content-Length"] = new[] { "0" };
```

Comment:
I thought that would prevent "Connection: close" from being automatically sent, but that's not the case. I'll change this to be written all the time.

BTW Do you think we should store `new[] { "0" }` in a static field?